### ISSUE TITLE
Does not work with the latest empyrion builds.

### ISSUE THREAD

**Issue by NovaSki11z** [Riebart/EmpyrionGSTools#4]
Unable to read beyond the end of the stream

That is the error I get no matter how I try to make a bp from a .stl file. I know its not very likely that you are still working on this but if you could just see about a fix for this error it would be amazing. This has been working for years and now its broke there is something to say for it maybe it will work for years again with just another fix.

**Comment by Riebart** [comment]
Worst case scenario, this is a whole new blueprint file format that we need to parse.

But it's possible that we can keep using old style blueprints, and then letting the game itself import and convert them.

**Comment by NovaSki11z** [comment]
I'm not really sure what you mean from what you are saying sounds like your
saying to put a STL in the game files. That's what I'm trying is to put an
STL into the egstools and turning it into a bp

**Comment by Riebart** [comment]
No. What I'm saying is that the file format Empyrion uses may have changed in a significant way, and the parsing may need to be updated.

I'll take a look at this next week.

**Comment by NovaSki11z** [comment]
Oh well thank you so much for all your work I tried seeing about doing it for you but man the amount of work in this code is insane lol I am not that good yet still got a lot to learn.

**Comment by NovaSki11z** [comment]
When you update this are you just releasing the update on this github? Just making sure not really rushing at all

**Comment by markLasagna** [comment]
It is partly still working. You only need a starting blueprint file in the old format. 
I think reading/parsing the ownerId and creatorId from the new blueprint format is enough to get it working.
I gues the old format will be supported for a long time because of the many existing blueprints.

**Comment by NovaSki11z** [comment]
No that is not true because I tried that and still it fails completely

**Comment by markLasagna** [comment]
It is working here right now. So what is not true? I used the original blueprint from the BlueprintBase folder. Using Empyrion 1.11.9 .

**Comment by davisn456** [comment]
Is there any chance of looking into this problem again? The current version of Empyrion doesn't seem to allow the blueprints to spawn anymore.

**Comment by AngelRed** [comment]
For me as well it's not working anymore, i tried creating blueprint in alpha 7, doing the magic with the tool to get it, going into 1.31.1 and jsut an internal error without having the possibility to spawn it.. I tried with a version that worked before for me, the 1.12, still nothing :c

**Comment by markLasagna** [comment]
Got it working, but currently working on tool from [ApanLoon](https://github.com/ApanLoon)
[EmpyrionStuff](https://github.com/ApanLoon/EmpyrionStuff)
ApanLoon already updated and expanded Empyrion file usage and added UI until year 2020.

I added voxelizer code from Riebart to import STL files. Replaced smoothing with 'find block-shape to follow STL surface' 

Got Riebart code working on current Emperion before I found EmpyrionStuff. So have to combine these someday to update EmpyrionStuff file handling.

Never used GitHub. Don't know about permissions on using or updating code from others.

<img width="1079" height="840" alt="Image" src="https://github.com/user-attachments/assets/41764bdb-fcf0-4171-aa9b-d2828478944b" />

**Comment by davisn456** [comment]
This is excellent news! If you're willing to share the working version of the STL importer, I'd love to try it out. I've really missed having this tool available.

**Comment by markLasagna** [comment]
https://codeberg.org/Drijver/EmpyrionStuff/releases
I'm new to this. Any comments are welcome.

**Comment by markLasagna** [comment]
Thank you for your reply.
Did you get a chance to try the import function?
I'm curious if you can use it, mostly because the interface is so different
from Riebart's tools.

**Comment by davisn456** [comment]
I selected a STL file to import, then set a size (50). I left the rest of the settings alone. When I started it, the app seemed to just crash immediately with no error messages. Is there a log somewhere I can look at? I tried several STL files to make sure it wasn't just an issue with the first one I tried, and I got the same result.

**Comment by MoreBlueGreen** [comment]
I have been trying it, and the program will let me look at EPB. files, but
it closes itself every time I try to import an STL. file.

**Comment by markLasagna** [comment]
Thank you for the interest and feedback.
So strange that there was no exception handling and logging.
I updated the code and added a new zip-file to the same release.
This adds some exception handling and messages.

**Comment by davisn456** [comment]
Whatever you did made a big difference. I was able to import two separate STL files and spawned the blueprints in the game. BIG PROGRESS!!! Just out of curiosity, is it possible to include code to smooth out the model a little bit? The old software had a smoothing function that really helped to eliminate jagged edges on the models. Otherwise, I love it!

**Comment by ApanLoon** [comment]
@markLasagna Cool to see that my old code is still useful for someone. I had no idea that it could still be used. 😀

**Comment by markLasagna** [comment]
I can try to copy the smoothing created by Riebart. Already tried to convert the Phyton code. But that only works when you understand what the code does. I'll start on it.
The function I created to follow the surface can also be extended with smoothing. But that will take more time. 
@ApanLoon Thanks for all your work. If you like one of us could merge the code to your github project. Don't know what is normal in such cases. This is my first git project.

**Comment by ApanLoon** [comment]
I think that the normal procedure would be for you to make a pull request from your forks back to my repo.

**Comment by Kip7777** [comment]
@markLasagna I too had success with your version, 0.6.1. With time, I could certainly work the outputted BP into a usable vessel. Not sure what would take less work currently, manually smoothing out a BP produced with full cubes or fixing up a BP made with the smoothing. Either would take a lot of time, although I did use a rather organic looking ship as my test which probably wasn't a great idea.

Quick Edit: Just some feedback after using both Riebart's tool and yours. The smoothing in Riebart's works much better (without corners though) as it only uses about 3 blocks for the smoothing process, all fairly simple ramps, where as yours seems to pick all kinds of blocks which would be time consuming to correct and not all of them look air tight.

**Comment by Kip7777** [comment]
To get back on topic for @NovaSki11z , I did some playing and have figured out a (somewhat convoluted) work around but it only works for Steam players sadly, as Epic does not allow you pick and choose which game version you install. I have access to EGS on both Epic and Steam which made this much faster for me to experiment with. If you only have the Steam version, you could setup a VM and install the older version of EGS in that, in order to not disturb your main Steam install. 

Requires Empyrion version Alpha 8 - other versions may work but this is what I used. This guide assumes you are making a CV, I have not tested bases or SV's/HV's yet. How to: Open your Steam account > Library > right click the Empyrion Galactic Survival entry > Properties > select 'Game Versions & Betas' > Choose Alpha 8 (Dec 17th 2018) > let the game "update". You should then see a [alpha8] behind your game entry. 

**WARNING:** If you do this on your main Steam install, **DO NOT** load any newer save file that you care about inside an older version of the game as you may just lose it! Ignore warnings and internal errors related to newer save files and blueprints, just click continue if you see any. 

### Useful Empyrion commands:

```
di = toggle debug info on/off, doesn't show as much information in old versions of Empyrion

gm fly = toggle godmode with flying on/off

destroy <entity id> = destroys whatever base/vessel you're looking at, use with di on

replaceblocks <entity id> <old block> <new block> = replaces all of the first type of block with whatever block you specify, in the base/vessel you're looking at, use with di on

prefabinfo <blueprint name> = This will list various information about a blueprint including any forbidden blocks it contains. Your blueprint name needs to be a single word without spaces (and probably also no special characters).
```



### Instructions to make Riebart's tool play nicely with current versions of EGS:

- Open Alpha 8 and start a new Creative game (name it Alpha 8 so you don't open it accidentally in a later game version).

- Create a new CV with a starter block. Delete all blocks, including the core, except for 1 plain hull block and save as a blueprint, then quit the creative save (you do not need to close Empyrion).

- Copy your newly saved 1-block blueprint file into BlueprintBase directory of Riebart's tool, rename it to BlueprintBase.epb (the original can be deleted but your epb MUST be named correctly for the tool to work).

- Use Riebart's tool (ensure you have read and understood the readme) to convert your stl to an epb, make sure you pick CV type for a CV.

- Cut the new epb, go into your blueprint directory, add a new folder and paste the epb into it, make sure the file name and its folder name match eg. test.epb inside of test.

- Open your Alpha 8 creative game again, enable godmode and spawn in your new blueprint and add a core.

- Use a colour tool to copy a default colour palette to your new vessel and re-colour it white using the replace feature of the colour tool, this is to make inspection easier and to give you a full palette to work with later on.

- Have a good look around and check you are happy with it. Large models may have bits chopped off when spawned, if that happens you will need to try again at a smaller size. You can also play with rotation options in Riebart's tool if it didn't spawn on the correct axis.

- Riebart's tool, even with the CV option selected, will still use Steel Blocks S for the hull, which are SV/HV blocks, however, they will be the correct physical size for a CV. Take a multitool and check the hitpoints of a block, if it says 50/50, it's using small blocks. Steel Blocks L should be 500/500 hitpoints. We must fix this now in Alpha 8, otherwise you lose any smoothing done by Riebart's tool should you decide to do this later, for example in 1.16, where all blocks will become cubes. Use the replaceblocks command to change HullFull into HullFullLarge.

- Save as a new blueprint, this one should now spawn correctly in 1.16 with no errors.

- As a final step, spawn the new blueprint in 1.16. The text in the blueprint list should be white which indicates no issues, if it's red, use prefabinfo command to see what went wrong. After spawning it in, immediately save it and tick the box to clear the pivot point. Destroy the vessel and spawn the newest blueprint to begin working on it!
